### PR TITLE
manual attendance uplaod api

### PR DESCRIPTION
**Pull request by nodeexcel**

### CODE REVIEW

File: src/api/index.js [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
 import upload from "./upload.js";
+import fetch from "./manualAttendance.js"
 
 export default () => {
     let api = Router();
 
     api.route('/user/attendance/').get(upload.attendance);
     api.route('/attendance/update_time_by_employee').get(upload.attendanceByEmployee);
+    api.route('/attendance/manual/').get(fetch.manualUpdatedAttendance);
+    api.route('/attendance/approoval').get(fetch.approovalAction);
```
Comment (Rahul-Khera-Codes) [inline src/api/index.js]: spelling "approovalAction"

File: src/api/manualAttendance.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import model from '../middleware';
+
+module.exports = {
+
+    manualUpdatedAttendance: (req, res, next) => {
+        model.get_manual_attendance.manualUpdate(req, res).then(() => {}).catch((err) => {
```
Comment (Rahul-Khera-Codes) [inline src/api/manualAttendance.js]: rename model to db everywhere. this model is not making sense

Comment (Rahul-Khera-Codes) [reply src/api/manualAttendance.js]: and what is "get_manual_attendance" ?

File: src/middleware/attendance_approoval.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import db from "../db";
+
+module.exports = {
+
+    approoveUpdate: function(req, res) {
```
Comment (Rahul-Khera-Codes) [inline src/middleware/attendance_approoval.js]: why is this a middleware? the purpose of middleware is something different

File: src/api/upload.js [hunk 4]
```diff
@@ -1,15 +1,17 @@
...
 module.exports = {
 
     attendance: (req, res, next) => {
-        model.upload.uploadAttendance(req, res).then((status) => {
+        upload.uploadAttendance(req, res).then((status) => {
             res.json({ message: status })
-        });
+        }).catch((err) => {
+            res.json(err);
+        })
```
Comment (Rahul-Khera-Codes) [inline src/api/upload.js]: all 3 above files are inside api/ folder this includes routes and logic as well.

what is the actual purpose of api/ folder it not clear to me. please explain

File: src/models/attendance_approval.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import db from "../db";
+
+module.exports = {
+
+    approveUpdate: function(req, res) {
```
Comment (Rahul-Khera-Codes) [inline src/models/attendance_approval.js]: if this is a model where is the table defination etc ?

File: src/controller/manualAttendance.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import db from '../db';
+
+module.exports = {
+
+    manualUpdatedAttendance: (req, res, next) => {
+        db.manual.manualUpdate(req, res).then((data) => { // gets all data manually entered by employees which is not reviewed
+            res.json(data)
+        }).catch((err) => {
```
Comment (Rahul-Khera-Codes) [inline src/controller/manualAttendance.js]: write this in short code everywhere

.catch((err) => {
 +            res.json(err);
 +        })

to

.catch(err =>res.json(err))

everywhere such small expression should be written in short code

File: src/models/attendance.js [hunk 7]
```diff
@@ -8,5 +14,89 @@ export default function(sequelize, DataTypes) {
         freezeTableName: true
     });
 
+    details.uploadAttendance = function(req, res) {
+            return new Promise((resolve, reject) => {
+                var oldpath = config.path;
+                var newpath = path.join(__dirname, './uploads/AGL_001.TXT');
+                fs.rename(oldpath, newpath, function(err) { //stores file to new path in upload folder
```
Comment (Rahul-Khera-Codes) [inline src/models/attendance.js]: still don't understand this rename. is this working on dev server?

File: src/models/attendance.js [hunk 8]
```diff
@@ -8,5 +14,89 @@ export default function(sequelize, DataTypes) {
...
+                    else {
+                        var lineReader = require('readline').createInterface({
+                            input: fs.createReadStream(newpath) //read file from new path in upload folder
+                        });
+                        var count = 0;
+                        var line_array = [];
+                        var filtered_data = [];
+
+                        function readFile(callback) {
+                            lineReader.on('line', function(line) {
```
Comment (Rahul-Khera-Codes) [inline src/models/attendance.js]: divide this function in smaller functions. and create helper classes. if not clear ask

File: src/models/attendance.js [hunk 9]
```diff
@@ -8,5 +14,89 @@ export default function(sequelize, DataTypes) {
...
+                details.findAll({ where: { user_id: req.query.userid } }).then((data) => { //fetching all record for user id
+                    var errorCode = 0;
+                    if (data[0]) {
+                        _.forEach(data, function(employee) {
+                            if (employee.timing == entryTime || employee.timing == exitTime) { //comparing each fileterd entry with input time for entry and exit
+                                errorCode = 1;
+                                reject({ "error": errorCode, "data": { "message": "Record exists" } })
+                            }
+                        })
+                        if (errorCode == 0) { //checking for error code
```
Comment (Rahul-Khera-Codes) [inline src/models/attendance.js]: divide this in smaller function and use helpers

File: src/models/manualAttendance.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+                resolve(data)
+            }).catch((err) => {
+                reject(err)
+            })
+
+        })
+    }
+
+    manual.approveUpdate = function(req, res, db) {
+        return new Promise((resolve, reject) => {
```
Comment (Rahul-Khera-Codes) [inline src/models/manualAttendance.js]: this promise is not needed. u can simply do manual.update() ask me if not clear

File: src/models/uploads/AGL_001.TXT [hunk 11]
```diff
@@ -0,0 +1,42 @@
+No	TMNo	EnNo	Name	INOUT	Mode	DateTime
+1	1	302		0	0	2017/07/10 20:31:57
```
Comment (nodeexcel) [inline src/models/uploads/AGL_001.TXT]: remove uploaded file once data is fetched

File: src/models/manualAttendance.js [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        timestamps: false,
+        freezeTableName: true
+    });
+
+    manual.manualUpdate = function(req, res) {
+        return new Promise((resolve, reject) => {
+            manual.findAll({ where: { action: null } }).then((data) => {
+                resolve(data)
+            }).catch(err => reject(err))
+
```
Comment (nodeexcel) [inline src/models/manualAttendance.js]: code formatting

File: src/models/attendance.js [hunk 13]
```diff
@@ -1,3 +1,10 @@
+import path from "path";
+import fs from "fs";
+import _ from 'lodash';
+import config from '../config.json'
+import moment from 'moment';
+import helper from '../helper';
+
 export default function(sequelize, DataTypes) {
     var details = sequelize.define('attendance', { // inserting data to database of attendance
```
Comment (nodeexcel) [inline src/models/attendance.js]: use let instead of var

File: src/controller/manualAttendance.js [hunk 14]
```diff
@@ -0,0 +1,17 @@
+import db from '../db';
+
+module.exports = {
+
+    manualUpdatedAttendance: (req, res, next) => {
+        db.manual_attendance.manualUpdateAttendance(req, res).then((data) => { // gets all data manually entered by employees which is not reviewed
```
Comment (nodeexcel) [inline src/controller/manualAttendance.js]: response is always send through controller or a common response handler for all responses

File: src/controller/manualAttendance.js [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+            res.json(data)
+        }).catch(err => res.json(err))
+    },
+
+    approvalAction: (req, res, next) => {
+        db.manual_attendance.approveUpdatedAttendance(req, res, db, function(status) { //approves or decline the manual attendance request
```
Comment (nodeexcel) [inline src/controller/manualAttendance.js]: i think here also you can create promise

File: src/models/manualAttendance.js [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+                if (data) {
+                    if (data.action == true) {
+                        db.attendance.create({ user_id: data.user_id, timing: data.timing }).then(function() {
+                            callback(true);
+                        })
+                    } else {
+                        callback(false);
+                    }
+                } else {
+                    res.json({ error: config.errMsg1 })
```
Comment (nodeexcel) [inline src/models/manualAttendance.js]: use callback

File: src/controller/upload.js [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+    attendance: (req, res, next) => {
+        db.attendance.uploadAttendance().then((status) => {
+            res.json(status)
+        }).catch(err => res.json(err))
+    },
+
+    attendanceByEmployee: (req, res, next) => {
+        db.attendance.uploadAttendanceByUserId(req.query, db).then((status) => {
+            res.json(status)
+        }).catch(err => res.json(err))
```
Comment (nodeexcel) [inline src/controller/upload.js]: use error handler for error message

File: src/models/attendance.js [hunk 18]
```diff
@@ -8,5 +15,51 @@ export default function(sequelize, DataTypes) {
...
+                })
+            })
+        },
+
+        attendance.uploadAttendanceByUserId = (query, db) => {
+            return new Promise((resolve, reject) => {
+                helper.time.timeConvertion(query.date, query.entry_time, query.exit_time, function(entryTime, exitTime) { //converts date and time in required format
+                    attendance.findAll({ where: { user_id: query.userid } }).then((data) => { //fetching all record for user id
+                        let errorCode = 0;
+                        if (data[0]) {
```
Comment (nodeexcel) [inline src/models/attendance.js]: use data.length

File: src/models/manualAttendance.js [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+            manual_attendance.findAll({ where: { action: null } }).then((data) => { //fetch all entries in manual_attendance table which are not reviewed
+                resolve(data)
+            }).catch(err => reject(err))
+
+        })
+    }
+
+    manual_attendance.approveUpdatedAttendance = function(query, db, callback) {
+        manual_attendance.update({ action: query.action }, { where: { id: query.id } }).then(function() { //updtaes manual_attenadance table
+            manual_attendance.find({ where: { id: query.id } }).then(function(data) {
```
Comment (nodeexcel) [inline src/models/manualAttendance.js]: check data is updated or not

File: src/controller/manualAttendance.js [hunk 20]
```diff
@@ -0,0 +1,21 @@
+import db from '../db';
+
+module.exports = {
+
+    manualUpdatedAttendance: (req, res, next) => {
+        db.manual_attendance.manualUpdateAttendance().then((data) => { // gets all data manually entered by employees which is not reviewed
+            res.json({ data: data })
```
Comment (nodeexcel) [inline src/controller/manualAttendance.js]: send data as res.json({message:"", error:"", data:data})

File: src/controller/manualAttendance.js [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+        }).catch(err => next(err))
```
Comment (nodeexcel) [inline src/controller/manualAttendance.js]: error message is in proper format

**Pull request merged**